### PR TITLE
feat: add manual ci for querying grant role events

### LINKED ISSUE #337: (feat): add manual ci for querying grant role events

### PR DESCRIPTION
**Pull request by viraj124**
- closes #337

### CODE REVIEW

File: packages/solidity-contracts/scripts/hardhat/eventFilter.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        taskArgs.contract,
+        grantRoleEvenABI,
+        provider
+      );
+
+      const eventPayload: any = [];
+
+      try {
+        const events = await contract.queryFilter(
+          contract.filters.RoleGranted()
```
Comment (DefiCake) [inline packages/solidity-contracts/scripts/hardhat/eventFilter.ts:28]: What happens when you query too many blocks?

**Review by DefiCake — COMMENTED** [review]
I 'd say that instead of just logging the events, we pick every address that was ever granted a role, and then query hasRole, and filter those that currently have a role assigned. That way we build a summary of currently active roles in the contract.

Ideally we also can match the role IDs against well known roles that we have, like DEFAULT_ADMIN, PAUSER, COMMITTER... and log those role IDs that we do not recognise.

File: packages/solidity-contracts/scripts/hardhat/eventFilter.ts (hunk 1 above)
Comment (viraj124) [reply packages/solidity-contracts/scripts/hardhat/eventFilter.ts:28]: currently tested with a public rpc, triggering the ci frequently might result in rate limit issues, as currently we query all blocks.

With a private rpc we should be good, but I was also thinking if we can specify a block range too but for now decided to keep the number of inputs minimal, so didn't add that

Comment (PraneshASP) [reply packages/solidity-contracts/scripts/hardhat/eventFilter.ts:28]: imo, hardhat tasks are not the ideal way to do this. Also why do we need this to be in the ci?

Comment (viraj124) [reply packages/solidity-contracts/scripts/hardhat/eventFilter.ts:95]: we need to be able to manually trigger it and there is no tool with a ui live currently that we can use for filtering events, there is a rate limit error risk with the script approach but using a pvt rpc reduces that risk to a great extent

File: packages/solidity-contracts/scripts/hardhat/eventFilter.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        );
+
+        for (const event of events) {
```
Comment (DefiCake) [inline packages/solidity-contracts/scripts/hardhat/eventFilter.ts:98]: I think that you could filter the roles and addresses before hammering the RPC

File: packages/solidity-contracts/scripts/hardhat/eventFilter.ts (hunk 1 above)
Comment (DefiCake) [reply packages/solidity-contracts/scripts/hardhat/eventFilter.ts:95]: > imo, hardhat tasks are not the ideal way to do this. Also why do we need this to be in the ci?

Please propose alternatives. I am tempted to use external infra for this task, like envio.

Comment (PraneshASP) [reply packages/solidity-contracts/scripts/hardhat/eventFilter.ts:95]: I guess we can use the graph protocol or quickly spin up our own indexer using ponder.sh

File: packages/solidity-contracts/scripts/hardhat/eventFilter.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            eventArgs.role =
+              roles.find((role) => role.value === event.args[0])?.name ||
+              'UNKNOWN_ROLE';
+            eventArgs.account = event.args[1];
+
+            eventPayload.push(eventArgs);
+          }
+        }
+
+        writeFileSync('grantedRoles.json', JSON.stringify(eventPayload));
```
Comment (DefiCake) [inline packages/solidity-contracts/scripts/hardhat/eventFilter.ts:132]:
```suggestion
        writeFileSync(
          'grantedRoles.json',
          JSON.stringify(eventPayload, undefined, 2)
        );
```

Prettify the json

File: packages/solidity-contracts/scripts/hardhat/eventFilter.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+task(
+  'grant-role-event-filter',
+  'Filters grant role event for a specific contract to keep track of assigned roles'
+)
+  .addParam('contract', 'address of the contract')
+  .setAction(
+    async (taskArgs: any, hre: HardhatRuntimeEnvironment): Promise<void> => {
+      const provider = new hre.ethers.JsonRpcProvider(process.env.RPC_URL);
+
+      const grantRoleEvenABI = [
```
Comment (DefiCake) [inline packages/solidity-contracts/scripts/hardhat/eventFilter.ts:14]: Maybe use hre.ethers.getContractAt to get access to the typechain types

**Review by DefiCake — COMMENTED** [review]
The types are a bit all over the place, opened this in my IDE and saw a lot of red

File: packages/solidity-contracts/scripts/hardhat/eventFilter.ts (hunk 4 above)
Comment (viraj124) [reply packages/solidity-contracts/scripts/hardhat/eventFilter.ts:14]: yeah I looked into it but still got typing errors, seems like a typing bug, will use a fix similar to the one in `upgradeVerification` script

**Review by DefiCake — APPROVED** [review]

**Review by PraneshASP — APPROVED** [review]

**Pull request merged**